design/dhcpless_ipam.md

```diff
@@ -0,0 +1,472 @@
+
+#### Potential config-drive conflict on redeployment
+
+When a host is re-provisioned, there may be an existing `config-2` partition on the disk that is written as part of the
+deployment process, but not cleaned at the early pre-provisioning stage where network configuration is required.
+
+This could potentially result in the wrong configuration being read, or the firstboot tool may fail due to finding
+multiple config-drives.
+
+There are several potential workarounds for this situation, but it would be preferable for a common solution
+to be found which can benefit both Metal3 and the broader Ironic community,  as such solving this problem is considered
+outside the scope of this proposal, but we will follow and engage with the
+[Ironic community spec related to this problem](https://review.opendev.org/c/openstack/ironic-specs/+/906324/1/specs/approved/fix-vmedia-boot-config.rst )
+
+### Work Items
+
+* BMO PR adding support for new annotation and adjusting state-machine behavior
+* CAPM3 PR adding new resource and controller
+* Add e2e coverage
+
+### Dependencies
+
+#### Firstboot agent support
+
+An agent in the IPA ramdisk image is required to consume the network data provided via the processes outlined above.
+
+The Ironic DHCP-less documentation describes using glean (a minimal python-based cloud-init alternative), and community
+supported builds containing this tool [are available](https://artifactory.nordix.org/ui/repos/tree/General/metal3/images/ipa/staging/centos/9-stream)
+
+There are several other options such as cloud-init, or even custom scripts/tooling which may be coupled to the OS, so we
+do not define a specific solution as part of this proposal.
+
+### Test Plan
+
+* TODO - test coverage, is there any existing coverage for the IPAM flow we can extend?
+
+### Upgrade / Downgrade Strategy
+
+This will be a net-new API and controller so there should be no upgrade impact.
+
+Downgrade of environments leveraging the new API will not be possible, but given provisioning is not currently
+possible via this specific flow that seems unlikely to be an issue.
+
+### Version Skew Strategy
+
+N/A
+
+## Drawbacks
+
+### Not all disk formats are supported
+
+This approach will not work in some configurations, and specifically it won't work with `format: live-iso`
+since in that case Ironic doesn't build the virtualmedia image, so it can't inject any `network_data`
+
+Additionally Ironic may not support injecting `network_data` in every configuration, when the IPA image is
+supplied as kernel and initramfs combo then Ironic can convert it to an ISO and inject the node `network_data`
+but in the case where a pre-built ISO ramdisk ISO is provided.
+
+In practice this probably isn't a huge issue, since if the preprovisioning image workflow is used to generate
+a customized ISO, it can embed the configuration instead of Ironic to achieve the same result.
+
+### Network configuration failure can't be reflected in status
+
```

Comment:
I agree that this part is out of the scope of the proposal and it is not worse than when the IPA fails because of a DHCP configuration problem. But we should address the feedback given when the templater fails to generate the preprovisioningNetworkData for a given baremetal host. Feedback at the level of the template would be strange if the reason is specific to a BMH.  A condition on the BMH looks weird as this is not performed by the main controller. A solution would be to at least generate an event associated to the BMH.

Reply:
Thanks this is a good point - there are probably a few error-paths to consider:

- `baremetalhost.metal3.io/preprovisioningnetworkdata-reference` value does not reference a valid resource (perhaps BMO should validate this, but just check the resource exists?)
- Problem with the DataTemplate resource - the schema is enforced by the API but AFAICS for `Metal3DataTemplate` any other error will be reported via the resource status, so I think we'll do the same for `Metal3PreprovisioningDataTemplate` given they will likely share the same code
- Some other problem in CAPM3 related to generating and setting the Secret,  here we don't have a Machine resource during the pre-provisioning phase so will have to do as you suggest and generate an event associated with the BMH